File: Python/Robot.py
# This example shows a TimedRobot class that controls a robot's movement using two ctre.PWM motors. 

# The robotInit() method initializes the motors
# The autonomousPeriodic() and teleopPeriodic() methods run the autonomous and teleop code, respectively.

import wpilib
import wpilib.drive

class MyRobot(wpilib.TimedRobot):

    def robotInit(self):
        """
        This function is called upon program startup and
        should be used for any initialization code.
        """
        self.leftDrive = wpilib.PWMSparkMax(0)
        self.rightDrive = wpilib.PWMSparkMax(1)
        self.robotDrive = wpilib.drive.DifferentialDrive(
            self.leftDrive, self.rightDrive
        )
        self.controller = wpilib.XboxController(0)
        self.timer = wpilib.Timer()

        # We need to invert one side of the drivetrain so that positive voltages
        # result in both sides moving forward. Depending on how your robot's
        # gearbox is constructed, you might have to invert the left side instead.
        self.rightDrive.setInverted(True)

    def autonomousInit(self):
        """
        This function is called once each time the robot enters autonomous mode.
        """
        self.timer.restart()

    def autonomousPeriodic(self):
        """
        This function is called periodically during autonomous.
        """
        if self.timer.get() < 2.0:
            self.robotDrive.arcadeDrive(0.5, 0.0)
        else:
            self.robotDrive.stopMotor()

    def teleopInit(self):
        """
        This function is called once each time the robot enters teleoperated mode.
        """

    def teleopPeriodic(self):
        """
        This function is called periodically during teleoperated mode.
        """
        self.robotDrive.arcadeDrive(-self.controller.getLeftY(),
                                   -self.controller.getRightX())

    def testInit(self):
        """
        This function is called once each time the robot enters test mode.
        """

    def testPeriodic(self):
        """
        This function is called periodically during test mode.
        """
        pass

File: Cpp/Robot.cpp
#include <frc/TimedRobot.h>
#include <frc/drive/DifferentialDrive.h>
#include <frc/motorcontrol/MotorControllerGroup.h>
#include <frc/XboxController.h>

// Create a class for the robot's drivetrain
class RobotContainer {
public:
    RobotContainer() {
        // Initialize the left and right sides of the drivetrain
        leftMotor = new WPI_TalonFX(1);
        rightMotor = new WPI_TalonFX(2);

        // Create a motor controller group for the left and right sides
        leftDrive = new MotorControllerGroup(leftMotor);
        rightDrive = new MotorControllerGroup(rightMotor);

        // Create a differential drive object to control the robot's movement
        drivetrain = new DifferentialDrive(leftDrive, rightDrive);

        // Create an Xbox controller object to get input from the driver
        driverController = new XboxController(0);
    }

    // Set the left and right motor speeds
    void setMotorSpeeds(double leftSpeed, double rightSpeed) {
        leftMotor->Set(leftSpeed);
        rightMotor->Set(rightSpeed);
    }

private:
    // Motor controllers for the left and right sides of the drivetrain
    WPI_TalonFX* leftMotor;
    WPI_TalonFX* rightMotor;

    // Motor controller groups for the left and right sides of the drivetrain
    MotorControllerGroup* leftDrive;
    MotorControllerGroup* rightDrive;

    // Differential drive object to control the robot's movement
    DifferentialDrive* drivetrain;

    // Xbox controller object to get input from the driver
    XboxController* driverController;
};

class Robot : public frc::TimedRobot {
public:
    Robot() : frc::TimedRobot(10_ms) {} // Set the update rate to 10 ms

    void robotInit() override {
        // Initialize the robot container
        robotContainer = new RobotContainer();
    }

    void teleopPeriodic() override {
        // Get the left and right joystick values from the Xbox controller
        double leftJoystick = robotContainer->driverController->GetLeftY();
        double rightJoystick = robotContainer->driverController->GetRightY();


        // Set the motor speeds based on the joystick values
        robotContainer->setMotorSpeeds(leftJoystick, rightJoystick);
    }

private:
    // Robot container object to control the robot's subsystems
    RobotContainer* robotContainer;
};


// In this example, we create a RobotContainer class to control the robot's drivetrain. The RobotContainer constructor initializes the left and right motors, creates a motor controller group for each side, and creates a DifferentialDrive object to control the robot's movement. It also creates an XboxController object to get input from the driver.

// The Robot class extends TimedRobot and sets the update rate to 10 ms. In the robotInit method, we initialize the RobotContainer object. In the teleopPeriodic method, we get the left and right joystick values from the Xbox controller and set the motor speeds based on those values.

// This example uses a differential drive, where the left and right motors are used together to move the robot forward or backward. The DifferentialDrive object is used to control the robot's movement.

// Note that this example uses the WPILib C++ framework and requires a functioning Xbox controller connected to the robot's driver station.
